### PR TITLE
Apply reversed colorscale setting and coloraxis in contour and histogramcontour legends

### LINKED ISSUE #4354: Apply reversescale in contour and histogram2dcontour legends
Expecting `contour` legend to flip fill direction when `reversescale` is set.
[demo](https://codepen.io/MojtabaSamimi/pen/ExxdBJR?editors=0010)

### LINKED ISSUE #4355: Problem drawing legends of contour and histogram2dcontour when using coloraxis
See console error.
[contour demo](https://codepen.io/MojtabaSamimi/pen/abbQqxB?editors=0010)
[histogram2dcontour demo](https://codepen.io/MojtabaSamimi/pen/oNNQEre?editors=0010)

### PR DESCRIPTION
**Pull request by archmoj**
Fixes #4354 and fixes #4355 
- displays identical legends for identical graphs
- use `coloraxis` colorscale (if defined) to style `legend`
- use `baseAttrs` instead of `plotAttrs` in all trace/attributes.js files (refactor commit https://github.com/plotly/plotly.js/pull/4356/commits/0c00efc4a1f22ab770d8cb96e87f4f0703931433)

@alexcjohnson 
@plotly/plotly_js 
cc: @nicolaskruchten

### CODE REVIEW

File: src/components/legend/style.js [hunk 1]
```diff
@@ -104,6 +114,28 @@ module.exports = function style(s, gd) {
         var showGradientFill = false;
         var dMod, tMod;
 
+        var colorscale = getColorscale(trace);
+        var reversescale = getReversescale(trace);
```
Comment (etpinard) [inline src/components/legend/style.js]: Can we reuse `extractOpts` instead 

https://github.com/plotly/plotly.js/blob/8951847f7edecd674cfebd2b670b550a97396745/src/components/colorscale/helpers.js#L65-L106

Something like:

```js
var cOpts = extractOpts(trace);
var colorscale = cOpts.colorscale;
var reversescale = cOpts.reverscale;
```

Comment (etpinard) [reply src/components/legend/style.js]: ... and we wouldn't need `getColorscale` nor `getReversescale`

File: src/components/legend/style.js [hunk 2]
```diff
@@ -472,3 +487,7 @@ module.exports = function style(s, gd) {
         }
     }
 };
+
+function getGradientDirection(reversescale) {
```
Comment (etpinard) [inline src/components/legend/style.js]: Looks like `reversescale` is only used as an argument for this function. If that's the case, we should :hocho: this and have

```js
var cOpts = extractOpts(trace);
var colorscale = cOpts.colorscale;
var gradientDirection = cOpts.reversescale ? 'horizontal' : 'horizontalreversed';
```

Comment (archmoj) [reply src/components/legend/style.js]: True. But soon it would be used in other places when adding new legends 3-D.

Comment (etpinard) [reply src/components/legend/style.js]: `getGradientDirection` or `reversescale` will get used?

Comment (archmoj) [reply src/components/legend/style.js]: Both.

Comment (etpinard) [reply src/components/legend/style.js]: All right - if you say so. Closing this comment, but please add a commit for https://github.com/plotly/plotly.js/pull/4356#discussion_r347573960

File: src/components/legend/style.js (hunk 1 above)
Comment (archmoj) [reply src/components/legend/style.js]: Good call. Done in https://github.com/plotly/plotly.js/pull/4356/commits/f988a643517d388af38ccd9586a4ec2065eb711a.

Comment (etpinard) [reply src/components/legend/style.js]: Awesome thanks!

**Comment by etpinard** [comment]
:dancer: nicely done @archmoj (once that flaky test passes)

**Pull request merged**